backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CPUPolicyUnit.java

```diff
@@ -39,56 +42,74 @@ public CPUPolicyUnit(PolicyUnit policyUnit,
+
+        boolean anyVmWithUnsetDynamicTopology = vmGroup.stream()
+                .anyMatch(vm -> VmCpuCountHelper.isResizeAndPinPolicy(vm)
+                            && !VmCpuCountHelper.isDynamicCpuTopologySet(vm));
+
+        // the number of host's shared CPUs has to >= as the number of CPUs of any VM
+        int requestedMaxSharedCpuCount = vmGroup.stream()
+                .filter(vm -> !vm.getCpuPinningPolicy().isExclusivelyPinned())
+                .mapToInt(vm -> VmCpuCountHelper.isDynamicCpuTopologySet(vm)
+                        ? vm.getCurrentNumOfCpus(false)
+                                : vm.getNumOfCpus(false))
+                .max()
+                .orElse(0);
+
+        for (VDS host : hosts) {
+
+            // when the dynamic topology is not set, the host requires more then one core per socket
+            if (anyVmWithUnsetDynamicTopology && host.getCpuCores() / host.getCpuSockets() <= 1) {
+                messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__NOT_ENOUGH_CORES.toString());
+                log.debug(
+                        "Host '{}' has only one core per socket. Resize and pin requires more than one core per socket",
+                        host.getName());
                 continue;
             }
-            Integer cores = SlaValidator.getEffectiveCpuCores(vds, context.getCluster().getCountThreadsAsCores());
-            if (cores != null) {
-                int numOfCpus = VmCpuCountHelper.isDynamicCpuTopologySet(vm) ?
-                        vm.getCurrentNumOfCpus(false) : vm.getNumOfCpus(false);
-                if (vm.getCpuPinningPolicy() == CpuPinningPolicy.DEDICATED) {
-                    int futureCpus = context.getCluster().getCountThreadsAsCores() ? cores - vds.getVmsCoresCount() :
-                            cores - (int) Math.ceil(vds.getVmsCoresCount() / (vds.getCpuThreads() / vds.getCpuCores()));
-                    if (numOfCpus > futureCpus) {
-                        messageNotEnoughCores(vds, cores, vm, messages);
-                        continue;
-                    }
-                } else {
-                    int takenCpus = 0;
-                    // takenCpus are CPUs (threads), we should consider switching it to cores when necessary.
-                    takenCpus = context.getCluster().getCountThreadsAsCores() ?
-                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
-                            vdsCpuUnitPinningHelper.countTakenCores(vds);
-                    cores = cores - takenCpus;
-                    if (numOfCpus > cores) {
-                        messageNotEnoughCores(vds, cores, vm, messages);
-                        continue;
-                    }
-                }
 
+            Integer hostCpuCount = SlaValidator.getEffectiveCpuCores(host, context.getCluster().getCountThreadsAsCores());
+
+            if (hostCpuCount == null) {
+                candidates.add(host);
```

Comment:
it was this way before but this makes it more explicit and I wonder if we really want to add the host in this case - if I read it correctly it means that we didn't get a report from the host about its cpu topology, no?

Reply:
what made me leave it like this is that the behavior is intentional - it has its own unit test so there might be a scenario where it is actually useful

Reply:
@liranr23 do you remember why?

Reply:
I think you are thinking about the `CpuPinningPolicyUnit` (https://github.com/oVirt/ovirt-engine/commit/bf91cba3037b5a3c4cb326873a49e12a1ac48e13).
In the `CPUPolicyUnit` it is like that forever (since Gilad Chaplik introduced this).

Reply:
indeed, I had CpuPinningPolicyUnit in mind..
ok, seems the test was written later on by a different person and as part of introducing CDI so they didn't seem to focus on the correctness of the tests (see https://gerrit.ovirt.org/c/ovirt-engine/+/46164) - I don't think we should take the test too seriously. but it's unrelated to this patch so we can do that separately

Reply:
It is unrelated to this patch, lets keep it the way it used to be.